### PR TITLE
feat: change config validation order

### PR DESCRIPTION
**Pull request by DmitryAnansky**
## What/Why/How?

- Changes to config validation flow, linting configuration with resolved config
- Extended types mixed into other oas types

Changes in this PR also results in code changes for Project

Affected changes & tests => https://github.com/Redocly/redocly/pull/7855

Screenshot 2024-04-15 at 14 44 19

## Reference
Closes https://github.com/Redocly/redocly-cli/issues/1223

## Check yourself

- [x] Code changed? - Tested with redoc/reference-docs/workflows (internal)
- [x] All new/updated code is covered with tests
- [ ] New package installed? - Tested in different environments (browser/node)

## Security

- [x] Security impact of change has been considered
- [x] Code follows company security practices and guidelines

### CODE REVIEW

File: .changeset/tender-vans-cover.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+"@redocly/openapi-core": minor
+"@redocly/cli": minor
+---
+
+Changed command execution flow to resolve configuration extendedTypes and plugins before linting it.
```
Comment (lornajane) [inline .changeset/tender-vans-cover.md:6]:
```suggestion
Changed resolution process to include extendedTypes and plugins before linting.
```

File: packages/core/src/config/load.ts [hunk 2]
```diff
@@ -93,19 +94,38 @@ export async function loadConfig(
...
     customExtends,
     configPath,
     tokens,
     files,
     region,
   });
+
+  if (configPath) {
+    try {
+      const { document, resolvedRefMap } = await resolveConfigFileAndRefs({
```
[the remark this answers is not in the document]
Comment (DmitryAnansky) [reply packages/core/src/config/load.ts:124]: `configPath` is not mandatory, we will resolve the configuration even without it.
But when present additional resolution steps are required.

**Review by lornajane — APPROVED** [review]
changelog is fine

File: __tests__/check-config/config-type-extensions-in-assertions/openapi.yaml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+                    properties:
+                      foo:
+                        type: string
+                type: object
+                oneOf:
+                  - properties:
+                      wrapper:
+                        $ref: '#main_data'
+                  - $ref: '#main_data'
+              example:
```
Comment (tatomyr) [inline __tests__/check-config/config-type-extensions-in-assertions/openapi.yaml:31]: We don't need the `paths` here.

File: __tests__/check-config/config-type-extensions-in-assertions/openapi.yaml [hunk 4]
```diff
@@ -0,0 +1,32 @@
+openapi: 3.1.0
+info:
+  title: Food Empire API
+  version: 0.5.1
+  x-metadata:
```
Comment (tatomyr) [inline __tests__/check-config/config-type-extensions-in-assertions/openapi.yaml:5]: Could we also test if the linter finds a wrong property?

File: packages/core/src/config/__tests__/load.test.ts [hunk 5]
```diff
@@ -104,22 +186,16 @@ describe('findConfig', () => {
 describe('getConfig', () => {
   jest.spyOn(fs, 'hasOwnProperty').mockImplementation(() => false);
   it('should return empty object if there is no configPath and config file is not found', () => {
-    expect(getConfig()).toEqual(Promise.resolve({}));
+    expect(getConfig()).toEqual(Promise.resolve({ rawConfig: {} }));
```
Comment (tatomyr) [inline packages/core/src/config/__tests__/load.test.ts:189]: Not sure why have this changed and if it's ok...

File: packages/core/src/config/load.ts [hunk 6]
```diff
@@ -80,10 +80,12 @@ async function addConfigMetadata({
...
 
-export type RawConfigProcessor = (
-  rawConfig: Document,
-  resolvedRefMap: ResolvedRefMap
-) => void | Promise<void>;
+export type RawConfigProcessor = (params: {
+  document: Document;
+  resolvedRefMap: ResolvedRefMap;
+  config: Config;
+  bundledConfig: Document['parsed'];
```
Comment (tatomyr) [inline packages/core/src/config/load.ts:87]: Maybe rename to `parsed`?

File: packages/core/src/types/redocly-yaml.ts [hunk 7]
```diff
@@ -1057,14 +1050,21 @@ const ConfigMockServer: NodeType = {
   },
 };
 
-export const createConfigTypes = (extraSchemas: JSONSchema) => {
+export const createConfigTypes = (extraSchemas: JSONSchema, config?: Config) => {
```
Comment (tatomyr) [inline packages/core/src/types/redocly-yaml.ts:1053]: Please use function declaration instead.

**Review by tatomyr — COMMENTED** [review]
Mostly minor comments.

**Review by tatomyr — APPROVED** [review]

**Pull request merged**